File: src/test/java/fr/axl/lvy/order/OrderBServiceTest.java
package fr.axl.lvy.order;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import fr.axl.lvy.client.Client;
import fr.axl.lvy.client.ClientRepository;
import java.time.LocalDate;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
@Transactional
class OrderBServiceTest {

  @Autowired OrderBService orderBService;
  @Autowired OrderBRepository orderBRepository;
  @Autowired OrderARepository orderARepository;
  @Autowired ClientRepository clientRepository;

  private OrderA createOrderA(String number) {
    var client = clientRepository.save(new Client("CLI-OB-" + number, "Client"));
    var order = new OrderA(number, client, LocalDate.of(2026, 3, 1));
    return orderARepository.save(order);
  }

  private OrderB createOrderB(String number, OrderA orderA, OrderB.OrderBStatus status) {
    var order = new OrderB(number, orderA);
    order.setStatus(status);
    return orderBRepository.save(order);
  }

  @Test
  void save_and_retrieve_orderB() {
    var orderA = createOrderA("CA-OB-01");
    var orderB = new OrderB("CB-2026-0001", orderA);
    orderBService.save(orderB);

    var found = orderBService.findById(orderB.getId());
    assertThat(found).isPresent();
    assertThat(found.get().getOrderNumber()).isEqualTo("CB-2026-0001");
    assertThat(found.get().getOrderA().getId()).isEqualTo(orderA.getId());
  }

  @Test
  void soft_delete_excludes_from_findAll() {
    var orderA = createOrderA("CA-OB-02");
    var orderB = createOrderB("CB-DEL-01", orderA, OrderB.OrderBStatus.SENT);

    orderBService.delete(orderB.getId());
    orderBRepository.flush();

    assertThat(orderBService.findAll()).noneMatch(o -> o.getOrderNumber().equals("CB-DEL-01"));
  }

  @Test
  void status_transition_sent_to_confirmed() {
    var orderA = createOrderA("CA-OB-03");
    var orderB = createOrderB("CB-ST-01", orderA, OrderB.OrderBStatus.SENT);

    var updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.CONFIRMED);
    assertThat(updated.getStatus()).isEqualTo(OrderB.OrderBStatus.CONFIRMED);
  }

  @Test
  void status_transition_confirmed_to_in_production() {
    var orderA = createOrderA("CA-OB-04");
    var orderB = createOrderB("CB-ST-02", orderA, OrderB.OrderBStatus.CONFIRMED);

    var updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.IN_PRODUCTION);
    assertThat(updated.getStatus()).isEqualTo(OrderB.OrderBStatus.IN_PRODUCTION);
  }

  @Test
  void status_transition_in_production_to_received() {
    var orderA = createOrderA("CA-OB-05");
    var orderB = createOrderB("CB-ST-03", orderA, OrderB.OrderBStatus.IN_PRODUCTION);

    var updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.RECEIVED);
    assertThat(updated.getStatus()).isEqualTo(OrderB.OrderBStatus.RECEIVED);
  }

  @Test
  void status_transition_sent_to_cancelled() {
    var orderA = createOrderA("CA-OB-06");
    var orderB = createOrderB("CB-ST-04", orderA, OrderB.OrderBStatus.SENT);

    var updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.CANCELLED);
    assertThat(updated.getStatus()).isEqualTo(OrderB.OrderBStatus.CANCELLED);
  }

  @Test
  void invalid_status_transition_throws() {
    var orderA = createOrderA("CA-OB-07");
    var orderB = createOrderB("CB-ST-05", orderA, OrderB.OrderBStatus.SENT);

    assertThatThrownBy(() -> orderBService.changeStatus(orderB, OrderB.OrderBStatus.RECEIVED))
        .isInstanceOf(IllegalStateException.class);
  }

  @ParameterizedTest
  @EnumSource(
      value = OrderB.OrderBStatus.class,
      names = {"RECEIVED", "CANCELLED"})
  void no_transitions_from_terminal_statuses(OrderB.OrderBStatus terminal) {
    var orderA = createOrderA("CA-OB-T" + terminal.ordinal());
    var orderB = createOrderB("CB-TERM-" + terminal.ordinal(), orderA, terminal);

    assertThatThrownBy(() -> orderBService.changeStatus(orderB, OrderB.OrderBStatus.CONFIRMED))
        .isInstanceOf(IllegalStateException.class);
  }

  @Test
  void markReceived_sets_reception_data() {
    var orderA = createOrderA("CA-OB-08");
    var orderB = createOrderB("CB-REC-01", orderA, OrderB.OrderBStatus.IN_PRODUCTION);

    var received =
        orderBService.markReceived(orderB, LocalDate.of(2026, 4, 1), true, "All items conform");

    assertThat(received.getStatus()).isEqualTo(OrderB.OrderBStatus.RECEIVED);
    assertThat(received.getReceptionDate()).isEqualTo(LocalDate.of(2026, 4, 1));
    assertThat(received.getReceptionConforming()).isTrue();
    assertThat(received.getReceptionReserve()).isEqualTo("All items conform");
  }

  @Test
  void markReceived_with_non_conforming_reception() {
    var orderA = createOrderA("CA-OB-09");
    var orderB = createOrderB("CB-REC-02", orderA, OrderB.OrderBStatus.CONFIRMED);

    var received =
        orderBService.markReceived(orderB, LocalDate.of(2026, 4, 2), false, "3 items damaged");

    assertThat(received.getReceptionConforming()).isFalse();
    assertThat(received.getReceptionReserve()).isEqualTo("3 items damaged");
  }

  @Test
  void findByOrderAId_returns_linked_orders() {
    var orderA = createOrderA("CA-OB-10");
    createOrderB("CB-LINK-01", orderA, OrderB.OrderBStatus.SENT);

    var found = orderBRepository.findByOrderAId(orderA.getId());
    assertThat(found).hasSize(1);
    assertThat(found.get(0).getOrderNumber()).isEqualTo("CB-LINK-01");
  }

  @Test
  void default_status_is_SENT() {
    var orderA = createOrderA("CA-OB-11");
    var orderB = new OrderB("CB-DEF-01", orderA);
    assertThat(orderB.getStatus()).isEqualTo(OrderB.OrderBStatus.SENT);
  }
}

File: src/test/kotlin/fr/axl/lvy/order/OrderBServiceTest.kt
package fr.axl.lvy.order

import fr.axl.lvy.client.Client
import fr.axl.lvy.client.ClientRepository
import java.time.LocalDate
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
@Transactional
class OrderBServiceTest {

  @Autowired lateinit var orderBService: OrderBService
  @Autowired lateinit var orderBRepository: OrderBRepository
  @Autowired lateinit var orderARepository: OrderARepository
  @Autowired lateinit var clientRepository: ClientRepository

  private fun createOrderA(number: String): OrderA {
    val client = clientRepository.save(Client("CLI-OB-$number", "Client"))
    val order = OrderA(number, client, LocalDate.of(2026, 3, 1))
    return orderARepository.save(order)
  }

  private fun createOrderB(number: String, orderA: OrderA, status: OrderB.OrderBStatus): OrderB {
    val order = OrderB(number, orderA)
    order.status = status
    return orderBRepository.save(order)
  }

  @Test
  fun save_and_retrieve_orderB() {
    val orderA = createOrderA("CA-OB-01")
    val orderB = OrderB("CB-2026-0001", orderA)
    orderBService.save(orderB)

    val found = orderBService.findById(orderB.id!!)
    assertThat(found).isPresent
    assertThat(found.get().orderNumber).isEqualTo("CB-2026-0001")
    assertThat(found.get().orderA!!.id).isEqualTo(orderA.id)
  }

  @Test
  fun soft_delete_excludes_from_findAll() {
    val orderA = createOrderA("CA-OB-02")
    val orderB = createOrderB("CB-DEL-01", orderA, OrderB.OrderBStatus.SENT)

    orderBService.delete(orderB.id!!)
    orderBRepository.flush()

    assertThat(orderBService.findAll()).noneMatch { it.orderNumber == "CB-DEL-01" }
  }

  @Test
  fun status_transition_sent_to_confirmed() {
    val orderA = createOrderA("CA-OB-03")
    val orderB = createOrderB("CB-ST-01", orderA, OrderB.OrderBStatus.SENT)

    val updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.CONFIRMED)
    assertThat(updated.status).isEqualTo(OrderB.OrderBStatus.CONFIRMED)
  }

  @Test
  fun status_transition_confirmed_to_in_production() {
    val orderA = createOrderA("CA-OB-04")
    val orderB = createOrderB("CB-ST-02", orderA, OrderB.OrderBStatus.CONFIRMED)

    val updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.IN_PRODUCTION)
    assertThat(updated.status).isEqualTo(OrderB.OrderBStatus.IN_PRODUCTION)
  }

  @Test
  fun status_transition_in_production_to_received() {
    val orderA = createOrderA("CA-OB-05")
    val orderB = createOrderB("CB-ST-03", orderA, OrderB.OrderBStatus.IN_PRODUCTION)

    val updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.RECEIVED)
    assertThat(updated.status).isEqualTo(OrderB.OrderBStatus.RECEIVED)
  }

  @Test
  fun status_transition_sent_to_cancelled() {
    val orderA = createOrderA("CA-OB-06")
    val orderB = createOrderB("CB-ST-04", orderA, OrderB.OrderBStatus.SENT)

    val updated = orderBService.changeStatus(orderB, OrderB.OrderBStatus.CANCELLED)
    assertThat(updated.status).isEqualTo(OrderB.OrderBStatus.CANCELLED)
  }

  @Test
  fun invalid_status_transition_throws() {
    val orderA = createOrderA("CA-OB-07")
    val orderB = createOrderB("CB-ST-05", orderA, OrderB.OrderBStatus.SENT)

    assertThatThrownBy { orderBService.changeStatus(orderB, OrderB.OrderBStatus.RECEIVED) }
      .isInstanceOf(IllegalStateException::class.java)
  }

  @ParameterizedTest
  @EnumSource(value = OrderB.OrderBStatus::class, names = ["RECEIVED", "CANCELLED"])
  fun no_transitions_from_terminal_statuses(terminal: OrderB.OrderBStatus) {
    val orderA = createOrderA("CA-OB-T${terminal.ordinal}")
    val orderB = createOrderB("CB-TERM-${terminal.ordinal}", orderA, terminal)

    assertThatThrownBy { orderBService.changeStatus(orderB, OrderB.OrderBStatus.CONFIRMED) }
      .isInstanceOf(IllegalStateException::class.java)
  }

  @Test
  fun markReceived_sets_reception_data() {
    val orderA = createOrderA("CA-OB-08")
    val orderB = createOrderB("CB-REC-01", orderA, OrderB.OrderBStatus.IN_PRODUCTION)

    val received =
      orderBService.markReceived(orderB, LocalDate.of(2026, 4, 1), true, "All items conform")

    assertThat(received.status).isEqualTo(OrderB.OrderBStatus.RECEIVED)
    assertThat(received.receptionDate).isEqualTo(LocalDate.of(2026, 4, 1))
    assertThat(received.receptionConforming).isTrue
    assertThat(received.receptionReserve).isEqualTo("All items conform")
  }

  @Test
  fun markReceived_with_non_conforming_reception() {
    val orderA = createOrderA("CA-OB-09")
    val orderB = createOrderB("CB-REC-02", orderA, OrderB.OrderBStatus.CONFIRMED)

    val received =
      orderBService.markReceived(orderB, LocalDate.of(2026, 4, 2), false, "3 items damaged")

    assertThat(received.receptionConforming).isFalse
    assertThat(received.receptionReserve).isEqualTo("3 items damaged")
  }

  @Test
  fun findByOrderAId_returns_linked_orders() {
    val orderA = createOrderA("CA-OB-10")
    createOrderB("CB-LINK-01", orderA, OrderB.OrderBStatus.SENT)

    val found = orderBRepository.findByOrderAId(orderA.id!!)
    assertThat(found).hasSize(1)
    assertThat(found[0].orderNumber).isEqualTo("CB-LINK-01")
  }

  @Test
  fun default_status_is_SENT() {
    val orderA = createOrderA("CA-OB-11")
    val orderB = OrderB("CB-DEF-01", orderA)
    assertThat(orderB.status).isEqualTo(OrderB.OrderBStatus.SENT)
  }
}
